File: src/main/kotlin/fiftyPopularCodintInterviewProblems/ReverseLinkedList.java
package fiftyPopularCodintInterviewProblems;


import java.util.ArrayList;
import java.util.List;

public class ReverseLinkedList {
    static class Node{
        int data;
        Node next;
        Node(int data){this.data = data; this.next = null;}
        Node(int data, Node next){this.data = data; this.next = next;}
    }

    static class LinkedList{
        Node head;
        LinkedList(){this.head = null;}
        LinkedList(Node head){this.head = head;}
    }

    public void reverseList(LinkedList list){
        Node previous = null;
        Node currentNode = list.head;
        while (currentNode != null) {
            Node next = currentNode.next;
            currentNode.next = previous;
            previous = currentNode;
            currentNode = next;
        }
        list.head = previous;
    }
}

File: src/main/kotlin/leetCode/medium/lincked_list/ReverseLinkedList.kt
package leetCode.medium.lincked_list


class ReverseLinkedList {

    class ListNode(var value: Int, var next: ListNode? = null) {
    }

    fun reverseList(): ListNode? {
        var head: ListNode? = ListNode(1, ListNode(2, ListNode(3)))

        var p: ListNode? = null
        var c: ListNode? = head

        while (c != null) {
            val temp = c.next
            c.next = p
            p = c
            c = temp
        }
        return c
    }

    fun reverseListRecursionally(): ListNode? {
        var head: ListNode? = ListNode(1)
        var node = recursion(head)
        while (node != null) {
            print("${node.value}")
            node = node.next
        }
        return node
    }

    fun recursion(list:ListNode?): ListNode? {
        if (list?.next?.next == null) {
            list?.next?.next = list
            val temp = list?.next
            list?.next = null
            return temp
        }
        val node = recursion(list.next)
        list.next?.next = list
        list.next = null
        return node
    }

}